### PR TITLE
Fix iOS last up event velocity calculation

### PR DESCRIPTION
**Pull request by xster**
Before
giphy-downsized-large 1

After
giphy-downsized-large

Remove the last gesture up event that got synthesized as a move event in the gesture converter from iOS velocity calculation. Doesn't affect Android.

Make the original platform's gesture event type accessible in the PointerEvent.

Doesn't make iOS scrolling perfectly match the platform's yet but fixes the most noticeable momentum loss problem that makes scrolling ~1/3 incorrect. 

#10977

### CODE REVIEW

File: packages/flutter/lib/src/gestures/monodrag.dart [hunk 1]
```diff
@@ -127,7 +127,10 @@ abstract class DragGestureRecognizer extends OneSequenceGestureRecognizer {
     if (event is PointerMoveEvent) {
       final VelocityTracker tracker = _velocityTrackers[event.pointer];
       assert(tracker != null);
-      tracker.addPosition(event.timeStamp, event.position);
+      // PointerMoveEvent from platform's gesture up events are positionally redundant
+      // vs the last PointerMoveEvent. Do not use in velocity calculation.
+      if (event.sourceEventType != PointerChange.up)
+        tracker.addPosition(event.timeStamp, event.position);
```
Comment (Hixie) [inline packages/flutter/lib/src/gestures/monodrag.dart]: This doesn't seem right. Why are the timestamps not enough to make this work?

Comment (xster) [inline packages/flutter/lib/src/gestures/monodrag.dart]: The final touchesEnded we get from the engine comes exactly one frame after the last touchesMoved we get (which are all one frame apart). I don't think we can't tell apart the touchesEnded event from the others by looking at the timestamp (visualizable in https://github.com/flutter/flutter/issues/10977#issuecomment-314879703 graphs).

Comment (Hixie) [inline packages/flutter/lib/src/gestures/monodrag.dart]: Just so I understand correctly, touchesEnded is what we turn into PointerMove+PointerUp in this scenario?

Comment (xster) [inline packages/flutter/lib/src/gestures/monodrag.dart]: If we originally got touchesMoved @ 0,0 -> touchesMoved @ 1,1 -> touchesEnded @ 1,1
we turned it into PointerMoveEvent @ 0,0 -> PointerMoveEvent @ 1,1 -> PointerMoveEvent @ 1,1. 

If I understand what you meant, ya, we turn the last touchesEnded into a PointerMoveEvent in class and up in sourceEventType

Comment (Hixie) [inline packages/flutter/lib/src/gestures/monodrag.dart]: what would the timestamps be on those pointer events?

Comment (xster) [inline packages/flutter/lib/src/gestures/monodrag.dart]: x, x+16, x+32

Comment (Hixie) [inline packages/flutter/lib/src/gestures/monodrag.dart]: so why would we want to drop the last one? Is it not "real"? Does the touchesEnded event randomly come much later? Does it come later with bogus data? I'm confused about what the event stream from the OS means here.

Comment (xster) [inline packages/flutter/lib/src/gestures/monodrag.dart]: ya, it's not really a move event as in the user still had the finger touching the screen at 1,1 at time x+32ms. The user stopped touching the screen at x+16ms. 

On Android, the sequence would have been PointerMoveEvent -> PointerMoveEvent -> PointerUpEvent which is semantically easier to parse than iOS's PointerMoveEvent -> PointerMoveEvent -> PointerMoveEvent where the last PointerMoveEvent was really a PointerUpEvent from the OS.

Comment (Hixie) [inline packages/flutter/lib/src/gestures/monodrag.dart]: So, two questions:
1) If the user stopped touching at x+16ms, why did the OS not tell us until x+32ms?
2) Does touchesEnded always give the same position as the last touchesMoved?

Comment (Hixie) [inline packages/flutter/lib/src/gestures/monodrag.dart]: Three questions:
3) What does the OS send us if the user does 0,0 at t=0ms, 1,1 at t=16ms, stays still for another 16ms, and releases (still at 1,1) at t=32ms?

Comment (xster) [inline packages/flutter/lib/src/gestures/monodrag.dart]: 1. The OS only sends gesture updates every 16ms as far as I can tell. 
2. It seems to always do +/- 10 pixels. Android always gives the exact same positions to the pixel. 
3. touchesMoved 0,0; touchesMoved 1,1; touchesMoved 1,1; touchesEnded 1,1 (if we omit the touchesStarted part). That gets transformed into PointerMoveEvent 0,0; PointerMoveEvent 1,1; PointerMoveEvent 1,1; PointerMoveEvent 1,1 for iOS in the framework.

Comment (Hixie) [inline packages/flutter/lib/src/gestures/monodrag.dart]: So what is the value of the coordinate in the last touchesEnded? (On iOS in general, I mean, not in Flutter.)

Comment (xster) [inline packages/flutter/lib/src/gestures/monodrag.dart]: Not much I think. Likely just to keep the UITouch object consistently filled across the different APIs. Android does so more or less too with MotionEvent having the coordinates on up.

**Comment by eseidelGoogle** [comment]
See also https://github.com/flutter/flutter/issues/5361#issuecomment-241186466

File: packages/flutter/lib/src/gestures/monodrag.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter/lib/src/gestures/monodrag.dart]: In that case it sounds like the real bug here is in the engine. We shouldn't be generating a move packet for the touchesEnd event.

**Comment by xster** [comment]
The coalescedTouchesForTouch API had a really big potential (not for this bug which can't be fixed with higher resolution but for as an alternative for the previous fix #1928 to enforce the same location for last move and end which causes this as a side effect). 

Unfortunately with testing, that API didn't produce additional data that helped bridge the small gap between the last move and the end events' locations. 

There was also another UITouch method preciseLocationInView: instead of locationInView:. Unfortunately it didn't produce data to help solve #1928 either.

@Hixie merging in thread from https://github.com/flutter/flutter/pull/11345#discussion_r129080602. The PointerMoveEvent generated from touchesEnded was deliberate AFAICT to conform to our API's guarantee that the last PointerMoveEvent will be exactly at the same location as PointerUpEvent. 

To keep that guarantee, I think we'll ultimately have to do some variant of this PR and prevent 'downsampling' of signals either by adding a new signal to un-hide the conversion or during the conversion, emit 2 events, one move event not at the current timestamp but rewrite history at the previous frame and also emit an up event. 

But I worry that with the latter solution, we're adding more layers of semantic abstraction that might confuse the user.

**Comment by Hixie** [comment]
We should only do the fixup once, at the iOS->engine boundary.
If the coordinate given by iOS for the up (touchEnded) is out of date or unreliable, then we should just ignore that data and use the data from the last actual touchMove when creating the data for the up event.

I don't think we should be changing any code in the framework to do this. The whole point of the contract is that the framework doesn't have to think about any of this.

**Comment by Hixie** [comment]
@xster is this PR obsolete now?

**Comment by xster** [comment]
Ya, if you prefer the other solution, closing this one.

**Pull request closed without merging**